### PR TITLE
Tint editor extension

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
# Features

Add tint editor to the thang editor color tab. Can be accessed via `/editor/thang` and then clicking on the Color tab.
This tab now has two modes. The default mode is the original purpose which is to define color groups. But there is now a button that switches the page to a tint playground tab and mode.

The tint assignment tab allows testing out various color tint combinations, and saving them to the database.

This tab includes a second treema for editing the tints.
It also replaces the grid of colors with dynamically created tints from the treema. This allows designers to combine the various tints together on the thang instead of one at a time as per prior behavior.

Once tints have been defined in the treema, they can now also be saved to the database if the user is an admin.

Screen Shot 2019-05-23 at 11 05 14 PM

Features in the image:
 - two hair tints defined by treema
 - generated tint color presets
 - ugly save button
 - preview of selected tints (only one per group)
 - buttons for switching between modes or tabs along top.

### CODE REVIEW

File: app/views/editor/thang/ThangTypeColorsTabView.coffee [hunk 1]
```diff
@@ -122,31 +191,44 @@ module.exports = class ThangTypeColorsTabView extends CocoView
...
 
-  addColorTintGroup: (buttons, groupName, colors) ->
-    buttons.append($("<h3>#{groupName}</h3>"))
+  addColorTintGroup: (buttons, tintName, tints, index) ->
+    buttons.append($("<h3>#{tintName}</h3>"))
+    saveButton = $("<button>#{tintName}</button>")
+    buttons.append($('<button />', {
+      text: "Save '#{tintName}' Tints",
+      class: 'save-btn',
+      click: ((index) => () =>
```
Comment (AndrewJakubowicz) [inline app/views/editor/thang/ThangTypeColorsTabView.coffee]: This is a way to add a closure in coffeescript. Let's me create a save button for each tint index.

File: ozaria/site/api/tint.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import fetchJson from 'app/core/api/fetch-json'
```
Comment (AndrewJakubowicz) [inline ozaria/site/api/tint.js]: Only exposing the `PUT` interface as there isn't a need to support the creation or `POST` interface. The two initial tint groups can be made manually.

File: app/views/editor/thang/ThangTypeColorsTabView.coffee [hunk 3]
```diff
@@ -2,10 +2,15 @@ require('app/styles/editor/thang/colors_tab.sass')
...
 SpriteBuilder = require 'lib/sprites/SpriteBuilder'
-{hexToHSL} = require 'core/utils'
+{hexToHSL, hslToHex} = require 'core/utils'
 require 'lib/setupTreema'
 createjs = require 'lib/createjs-parts'
 initSlider = require 'lib/initSlider'
+tintApi = require('../../../../ozaria/site/api/tint')
+tintSchema = require 'app/schemas/models/tint.schema.js'
+
+COLORGROUPTAB = 'COLORGROUPTAB'
```
Comment (jmif) [inline app/views/editor/thang/ThangTypeColorsTabView.coffee]: Maybe `COLOR_GROUP_TAB`?  Little easier to read

File: app/views/editor/thang/ThangTypeColorsTabView.coffee (hunk 1 above)
Comment (jmif) [reply app/views/editor/thang/ThangTypeColorsTabView.coffee]: This is a great candidate for a comment in code :)

File: app/views/editor/thang/ThangTypeColorsTabView.coffee [hunk 4]
```diff
@@ -49,8 +72,55 @@ module.exports = class ThangTypeColorsTabView extends CocoView
...
+  onColorGroupTab: ->
+    @tintAssignments?.destroy()
+    @tab = COLORGROUPTAB
+    @render()
+
+  onTintAssignmentTab: ->
+    @tab = TINTTAB
+    @render()
+
+    fetch("/db/tint/all")
```
Comment (jmif) [inline app/views/editor/thang/ThangTypeColorsTabView.coffee]: Any reason to not put this in the tints API module?

**Comment by AndrewJakubowicz** [comment]
Thanks @jmif. I added your feedback.

**Pull request merged**